### DISCUSSION TITLE
Varying routing structure per subdomain

### DISCUSSION THREAD

**Discussion by bros4president** [nuxt/nuxt#16536] [Questions]
Hi! I've found myself in a very tricky situation for my level of experience, I'd really value and learn from your thoughts on this:

My nuxt app should serve requests to `*.domain.com`, where `*` is a subdomain that can be either of type A or type B.

Type A and type B subdomains follow a slightly different routing structure (like they would have their own `/pages` directory).
They share most of the components and they consume the same backend APIs.
For now.

They might diverge a lot in the future as each type of subdomain serve a different group of people with currently common, but possibly diverging needs. So in the long term I think the best option would be to have each subdomain type have its own Nuxt project and backend apis, and share a common library of components. Both Nuxt projects' ./output folders would live in the same server and a small piece of code would decide which handler to execute based on the subdomain.

But now, just starting to diverge, I want to keep things as simple as possible. I've been having a hard time figuring out how to use the same Nuxt3 project with varying routing structure. These are solutions I've come up with:

1. Having an empty `/pages` directory to force Vue Router. Define 2 arrays of routes (complete with path, name, component, etc), one for each type of subdomain, somewhere in the project, and create a plugin that executes `router.addRoute` for every route in the selected array based on the subdomain.

The problem with this is I get a lot of Vue Router warns about not finding certain path, even though the routing succeded in the end. Probably because some lookup happened before the addRoute calls. So it feels a little bit hacky and end up losing most of the perks of playing Nuxt by the rules, like definePageMeta. **Is a plugin the best place to implement this type of solution?** I've seen the `'pages:extend'` hook being recommended in issues, but I don't have access to the subdomain at build time.

2. Have 2 directories: `/pages-A` and  `/pages-B`. Write a script that builds the project with the pages directory set to `/pages-A` and then do the same with `/pages-B`. Deploy both `./output` directories and write a small middleware outside Nuxt that picks the handler based on the subdomain.

Not really a problem here, except it seems a bit convoluted for such a simple use case. I'd like to keep things simple for now.

3. Save the type of subdomain in some form of global state and have the project polluted with if statements in components, plugins and route middlewares.

Thank you very much for your time :)! Would love to discuss other solutions to routing or different approaches to this use case 🙏🏻

**Comment by TheAlexLichter** [comment]
Check out https://github.com/zernonia/keypress/blob/main/app/router.options.ts for an option to handle ☺️
